pkgs/applications/networking/browsers/firefox/wrapper.nix

```diff
@@ -75,6 +83,88 @@ let
             ++ lib.optional (config.pulseaudio or true) libpulseaudio;
       gtk_modules = [ libcanberra-gtk2 ];
 
+      #########################
+      #                       #
+      #   EXTRA PREF CHANGES  #
+      #                       #
+      #########################
+      policiesJson = builtins.toFile "policies.json"
+        (builtins.toJSON enterprisePolicies);
+
+      extensions = builtins.map (a:
+        if ! (builtins.hasAttr "name" a) || ! (builtins.isString a.name) then
+          throw "Firefox addon needs a name attribute"
+        else if ! (builtins.hasAttr "url" a) || ! (builtins.isString a.url) then
+          throw "Addon ${a.pname} needs an url"
+        else if ! (builtins.hasAttr "sha256" a) || ! (builtins.isString a.sha256) then
+          throw "Addon ${a.pname} needs an sha256 checksum"
+        else stdenv.mkDerivation rec {
+          pname = a.name;
+          version = "1.0";
+          src = fetchurl {
```

Comment:
Instead of passing in data `name`, `url` and `sha256` I would just take a path and move the `fetchurl` outside of this. This makes it more flexible because one could also download xpi files and pass them diretory or use other ways of downloading the xpi.
Also I would externalize this builder into something like `buildFirefoxAddon` so that people can override build phases and apply patches.

Reply:
@rycee  Already has build an abstraction for that in `NUR`: buildFirefoxXpiAddon
It also has an mechanism for updating firefox addons.

Reply:
Hmm I get your point but there is one problem the extid of the addon has to change every time the hash changes if this is not done firefox does not update the addons in every profile, if we supply only the path then it is not possible to generate the extid because we don't have the checksum.

Reply:
> @rycee Already has build an abstraction for that in `NUR`: [buildFirefoxXpiAddon](https://github.com/nix-community/nur-combined/blob/2ae8d6629942a9bcd42780ca9a834c8758af32df/repos/rycee/pkgs/firefox-addons/default.nix#L5)
> It also has an mechanism for updating firefox addons.

Ahh cool but that needs some adjustments if you want it to work with every addon. If you look at my pull request in line 119,
I change two possible ids that have to be set, (one is the old way that is there for compatibility)